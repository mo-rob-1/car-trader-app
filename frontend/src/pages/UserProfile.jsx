import { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
// import { useParams, useNavigate } from "react-router-dom";
import { Link } from "react-router-dom";
import { toast } from "react-toastify";
import { getUserCars, updateCar, deleteCar, reset } from "../features/cars/carSlice";

function UserProfile() {
  const { cars, isLoading, isSuccess } = useSelector((state) => state.car);
  const { user } = useSelector((state) => state.auth);

  // const params = useParams();
  // const { carId } = useParams();

  const dispatch = useDispatch();
  // const navigate = useNavigate();

  useEffect(() => {
    return () => {
      if (isSuccess) {
        dispatch(reset());
      }
    };
  }, [isSuccess, dispatch]);

  useEffect(() => {
    dispatch(getUserCars());
  }, [dispatch]);

  if (isLoading) {
    return <h3 className="text-2xl font-bold">Loading...</h3>;
  }

  return (
    <>
      <div>
        <h1 className="text-3xl font-bold">My Profile</h1>
        <h2 className="text-2xl font-bold">Name: {user.name}</h2>
        <h3 className="text-2xl font-bold">Email: {user.email}</h3>
      </div>
      <h1 className="text-3xl font-bold">{user.name}'s Cars</h1>
      {cars.length > 0 ? (
        <div>
          {cars.map((car) => (
            <div className="card w-96 bg-base-100 shadow-xl my-5" key={car._id}>
              <figure>
                {" "}
                <img src={car.imageURL} alt={car.model} className="w-full" />
              </figure>
              <div className="card-body">
                <h2 className="card-title">{car.model}</h2>
                <p>Desc: {car.description}</p>
                <p>Price: Â£{car.price}</p>
                <p>Location: {car.location}</p>
                <p>Phone: {car.phoneNumber}</p>
                <p>Email: {car.email}</p>

                {car.status === "Available" ? (
                  <div className="bg-green-600 w-fit px-5 py-3 rounded-md my-3">
                    <p className="text-white font-bold">Available</p>
                  </div>
                ) : (
                  <div className="bg-red-600 w-fit px-5 py-3 rounded-md my-3">
                    <p className="text-white font-bold">Sold</p>
                  </div>
                )}

                <Link to={`/cars/${car._id}`} className="btn">
                  View
                </Link>
                <button
                  onClick={() => {
                    dispatch(updateCar(car._id, car.status));
                    toast.success("Car status updated");
                  }}
                  className="btn btn-success"
                >
                  Mark As Sold
                </button>

                <button onClick={() => dispatch(deleteCar(car._id))} className="btn btn-error">
                  Delete
                </button>
              </div>
            </div>
          ))}
        </div>
      ) : (
        <h3 className="text-3xl font-bold">No Cars Added.</h3>
      )}
    </>
  );
}

export default UserProfile;
