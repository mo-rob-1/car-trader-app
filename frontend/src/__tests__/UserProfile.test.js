import { render, screen, waitForDomChange } from "@testing-library/react";
import UserProfile from "../pages/UserProfile";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { store } from "../app/store";
import { BrowserRouter as Router } from "react-router-dom";

describe("UserProfile", () => {
  it("renders the user profile page", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <UserProfile />
        </Router>
      </Provider>
    );
    waitForDomChange().then(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("renders the My Profile text", () => {
    render(
      <Provider store={store}>
        <Router>
          <UserProfile />
        </Router>
      </Provider>
    );
    waitForDomChange().then(() => {
      expect(screen.getByText("My Profile")).toBeInTheDocument();
    });
  });

  it("renders the name of user and their email address", () => {
    render(
      <Provider store={store}>
        <Router>
          <UserProfile />
        </Router>
      </Provider>
    );
    window.fetch = jest.fn();
    window.fetch.mockResolvedValueOnce({
      json: async () => [{ name: "Mo", email: "mo@email.com" }],
    });
  });

  it("renders the text if no cars added yet", () => {
    render(
      <Provider store={store}>
        <Router>
          <UserProfile />
        </Router>
      </Provider>
    );
    waitForDomChange().then(() => {
      const { getComputedStyle } = window;
      window.getComputedStyle = (elt) => getComputedStyle(elt);
      screen.getByRole("heading", {
        name: /\- no cars added yet\./i,
      });
    });
  });
});
