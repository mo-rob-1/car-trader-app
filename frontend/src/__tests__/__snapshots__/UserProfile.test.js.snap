// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile renders the user profile page 1`] = `
<DocumentFragment>
  <div
    class="bg-base-200"
  >
    <div
      class="m-auto w-11/12 max-w-screen-2xl pt-5 pb-5 md:pt-12 md:pb-12"
    >
      <div
        class="bg-white p-5 md:p-8 rounded-2xl"
      >
        <h1
          class="text-3xl font-bold mb-5"
        >
          My Profile
        </h1>
        <hr
          class="mb-5"
        />
        <h2
          class="text-2xl font-bold flex items-center"
        >
          <svg
            class="mr-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
           
        </h2>
        <h3
          class="text-2xl font-bold flex items-center mt-4"
        >
          <svg
            class="mr-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M502.3 190.8c3.9-3.1 9.7-.2 9.7 4.7V400c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V195.6c0-5 5.7-7.8 9.7-4.7 22.4 17.4 52.1 39.5 154.1 113.6 21.1 15.4 56.7 47.8 92.2 47.6 35.7.3 72-32.8 92.3-47.6 102-74.1 131.6-96.3 154-113.7zM256 320c23.2.4 56.6-29.2 73.4-41.4 132.7-96.3 142.8-104.7 173.4-128.7 5.8-4.5 9.2-11.5 9.2-18.9v-19c0-26.5-21.5-48-48-48H48C21.5 64 0 85.5 0 112v19c0 7.4 3.4 14.3 9.2 18.9 30.6 23.9 40.7 32.4 173.4 128.7 16.8 12.2 50.2 41.8 73.4 41.4z"
            />
          </svg>
           
        </h3>
      </div>
      <h1
        class="text-3xl font-bold mt-10"
      >
        's Cars:
      </h1>
      <h3
        class="text-2xl font-bold mt-5"
      >
        - No cars added yet.
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
